// @fluencelabs/fluence/dist/internal/compilerSupport/v3impl/callFunction.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.callFunctionImpl = exports.callFunction = void 0;
var FluencePeer_1 = require("../../FluencePeer");
var index_1 = require("../../../index");
var services_1 = require("./services");
/**
 * Convenience function to support Aqua `func` generation backend
 * The compiler only need to generate a call the function and provide the corresponding definitions and the air script
 *
 * @param rawFnArgs - raw arguments passed by user to the generated function
 * @param def - function definition generated by the Aqua compiler
 * @param script - air script with function execution logic generated by the Aqua compiler
 */
function callFunction(rawFnArgs, def, script) {
    var _a = extractArgs(rawFnArgs, def), args = _a.args, peer = _a.peer, config = _a.config;
    return callFunctionImpl(def, script, config || {}, peer, args);
}
exports.callFunction = callFunction;
/**
 * Convenience function which does all the internal work of creating particles
 * and making necessary service registrations in order to support Aqua function calls
 *
 * @param def - function definition generated by the Aqua compiler
 * @param script - air script with function execution logic generated by the Aqua compiler
 * @param config - options to configure Aqua function execution
 * @param peer - Fluence Peer to invoke the function at
 * @param args - args in the form of JSON where each key corresponds to the name of the argument
 * @returns
 */
function callFunctionImpl(def, script, config, peer, args) {
    var argumentTypes = getArgumentTypes(def);
    var promise = new Promise(function (resolve, reject) {
        var e_1, _a;
        var particle = peer.internals.createNewParticle(script, config === null || config === void 0 ? void 0 : config.ttl);
        if (particle instanceof Error) {
            return reject(particle.message);
        }
        try {
            for (var _b = __values(Object.entries(args)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), name_1 = _d[0], argVal = _d[1];
                var type = argumentTypes[name_1];
                var service = void 0;
                if (type.tag === 'arrow') {
                    service = (0, services_1.userHandlerService)(def.names.callbackSrv, [name_1, type], argVal);
                }
                else {
                    service = (0, services_1.injectValueService)(def.names.getDataSrv, name_1, type, argVal);
                }
                (0, services_1.registerParticleScopeService)(peer, particle, service);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        (0, services_1.registerParticleScopeService)(peer, particle, (0, services_1.responseService)(def, resolve));
        (0, services_1.registerParticleScopeService)(peer, particle, (0, services_1.injectRelayService)(def, peer));
        (0, services_1.registerParticleScopeService)(peer, particle, (0, services_1.errorHandlingService)(def, reject));
        peer.internals.initiateParticle(particle, function (stage) {
            // If function is void, then it's completed when one of the two conditions is met:
            //  1. The particle is sent to the network (state 'sent')
            //  2. All CallRequests are executed, e.g., all variable loading and local function calls are completed (state 'localWorkDone')
            if (isReturnTypeVoid(def) && (stage.stage === 'sent' || stage.stage === 'localWorkDone')) {
                resolve(undefined);
            }
            if (stage.stage === 'sendingError') {
                reject("Could not send particle for ".concat(def.functionName, ": not connected  (particle id: ").concat(particle.id, ")"));
            }
            if (stage.stage === 'expired') {
                reject("Request timed out after ".concat(particle.ttl, " for ").concat(def.functionName, " (particle id: ").concat(particle.id, ")"));
            }
            if (stage.stage === 'interpreterError') {
                reject("Script interpretation failed for ".concat(def.functionName, ": ").concat(stage.errorMessage, "  (particle id: ").concat(particle.id, ")"));
            }
        });
    });
    return promise;
}
exports.callFunctionImpl = callFunctionImpl;
var isReturnTypeVoid = function (def) {
    if (def.arrow.codomain.tag === 'nil') {
        return true;
    }
    return def.arrow.codomain.items.length == 0;
};
/**
 * Arguments could be passed in one these configurations:
 * [...actualArgs]
 * [peer, ...actualArgs]
 * [...actualArgs, config]
 * [peer, ...actualArgs, config]
 *
 * This function select the appropriate configuration and returns
 * arguments in a structured way of: { peer, config, args }
 */
var extractArgs = function (args, def) {
    var argumentTypes = getArgumentTypes(def);
    var argumentNames = Object.keys(argumentTypes);
    var numberOfExpectedArgs = argumentNames.length;
    var peer;
    var structuredArgs;
    var config;
    if (FluencePeer_1.FluencePeer.isInstance(args[0])) {
        peer = args[0];
        structuredArgs = args.slice(1, numberOfExpectedArgs + 1);
        config = args[numberOfExpectedArgs + 1];
    }
    else {
        peer = index_1.Fluence.getPeer();
        structuredArgs = args.slice(0, numberOfExpectedArgs);
        config = args[numberOfExpectedArgs];
    }
    if (structuredArgs.length !== numberOfExpectedArgs) {
        throw new Error("Incorrect number of arguments. Expecting ".concat(numberOfExpectedArgs));
    }
    var argsRes = argumentNames.reduce(function (acc, name, index) {
        var _a;
        return (__assign(__assign({}, acc), (_a = {}, _a[name] = structuredArgs[index], _a)));
    }, {});
    return {
        peer: peer,
        config: config,
        args: argsRes,
    };
};
var getArgumentTypes = function (def) {
    if (def.arrow.domain.tag !== 'labeledProduct') {
        throw new Error('Should be impossible');
    }
    return def.arrow.domain.fields;
};
